src/ff/prime_field.rs

```diff
@@ -0,0 +1,307 @@
+                *self = *self + rhs;
+            }
+        }
+
+        impl Neg for $field {
+            type Output = Self;
+
+            fn neg(self) -> Self::Output {
+                Self((Self::PRIME - self.0) % Self::PRIME)
+            }
+        }
+
+        impl Sub for $field {
+            type Output = Self;
+
+            fn sub(self, rhs: Self) -> Self::Output {
+                // TODO(mt) - constant time?
+                let c = u128::from;
+                Self(
+                    ((c(Self::PRIME) + c(self.0) - c(rhs.0)) % c(Self::PRIME))
+                        as <Self as Field>::Integer,
+                )
+            }
+        }
+
+        impl SubAssign for $field {
+            #[allow(clippy::assign_op_pattern)]
+            fn sub_assign(&mut self, rhs: Self) {
+                *self = *self - rhs;
+            }
+        }
+
+        impl Mul for $field {
+            type Output = Self;
+
+            fn mul(self, rhs: Self) -> Self::Output {
+                // TODO(mt) - constant time?
+                let c = u128::from;
+                #[allow(clippy::cast_possible_truncation)]
+                Self(((c(self.0) * c(rhs.0)) % c(Self::PRIME)) as <Self as Field>::Integer)
+            }
+        }
+
+        impl MulAssign for $field {
+            #[allow(clippy::assign_op_pattern)]
+            fn mul_assign(&mut self, rhs: Self) {
+                *self = *self * rhs;
+            }
+        }
+
+        impl From<$field> for u8 {
+            fn from(v: $field) -> Self {
+                v.0
+            }
+        }
+
+        /// An infallible conversion from `u128` to this type.  This can be used to draw
+        /// a random value in the field.  This introduces bias into the final value
+        /// but for our purposes that bias is small provided that `2^128 >> PRIME`, which
+        /// is true provided that `PRIME` is kept to at most 64 bits in value.
+        ///
+        /// This method is simpler than rejection sampling for these small prime fields.
+        impl<T: Into<u128>> From<T> for $field {
+            fn from(v: T) -> Self {
+                #[allow(clippy::cast_possible_truncation)]
+                Self((v.into() % u128::from(Self::PRIME)) as <Self as Field>::Integer)
+            }
+        }
+
+        impl Debug for $field {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                write!(f, "{}_mod{}", self.0, Self::PRIME)
+            }
+        }
+    )* };
+}
+
+#[derive(Clone, Copy, PartialEq)]
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+pub struct Fp2(<Self as Field>::Integer);
+
+impl Field for Fp2 {
+    type Integer = u8;
+    const PRIME: Self::Integer = 2;
+    const ZERO: Self = Fp2(0);
+    const ONE: Self = Fp2(1);
+}
+
+impl BinaryField for Fp2 {}
+
+impl BitAnd for Fp2 {
+    type Output = Self;
+
+    fn bitand(self, rhs: Self) -> Self::Output {
+        Self((self.0 & rhs.0) & 1)
```

Comment:
This is because Fp2 is backed up by `u8` and implements `Field` with all arithmetic operations like Add, Mul. `& 1` is equivalent to `% Fp2::PRIME` and I thought `&` might perform slightly better. I'm not sure.

An example where this would be a problem is `Fp2(2) | Fp2(1)`, which will output Fp2(3) without `& 1` and fails this equality check: `assert_eq!(Fp2::ONE, Fp2(2) | Fp2(1))`.

Reply:
This is another symptom of the same root issue of not doing range checking in the Fp2::from method

Reply:
Fp31 clamps in it’s From impl